milli/src/update/index_documents/transform.rs

```diff
@@ -677,6 +774,39 @@ impl<'a, 'i> Transform<'a, 'i> {
     }
 }
 
+/// Merge all the obks in the order we see them.
+fn merge_obkvs_and_operations<'a>(_key: &[u8], obkvs: &[Cow<'a, [u8]>]) -> Result<Cow<'a, [u8]>> {
+    // [add, add, delete, add, add]
+    // we can ignore everything that happened before the last delete.
+    let starting_position = obkvs
+        .iter()
+        .rev()
+        .position(|obkv| obkv[0] == Operation::Deletion as u8)
+        .map_or(0, |pos| obkvs.len() - pos);
+
+    // [add, add, delete]
+    // if the last operation was a deletion then we simply return the deletion
+    if starting_position == obkvs.len() {
+        return Ok(obkvs[obkvs.len() - 1].clone());
+    }
+    let mut buffer = Vec::new();
+
+    // (add, add, delete) [add, add]
+    // in the other case, no deletion will be encountered during the merge
+    Ok(obkvs[starting_position..]
+        .iter()
+        .cloned()
+        .reduce(|acc, current| {
+            let first = obkv::KvReader::new(&acc[1..]);
+            let second = obkv::KvReader::new(&current[1..]);
+            merge_two_obkvs(first, second, &mut buffer);
+            // TODO: do this only once at the end
```

Comment:
What do you mean? Can you fix this TODO, please?

Reply:
I was saying that we don't need to insert this `Operation::Addition` at the start of the vector (which can be expensive if the vector gets big), but we could instead do it only once at the very end of the fold/reduce once we have our last obkv.

I've done it here; https://github.com/meilisearch/meilisearch/pull/3470/commits/1b1703a6090008ce7ebff5c5d9ed5c53fc0293e6